### PR TITLE
Add support for subscribing to StreamMessage with or without pooled o…

### PR DESCRIPTION
**Pull request by anuraaga**
…bjects.

Currently, it is possible for decorators to subscribe to a response and return a different one without releasing pooled objects. This makes the default behavior of subscription do a copy from pooled to unpooled to protect unknowing decorators from memory leaks. As modifying the response content in a decorator is quite rare, this copy should rarely happen in practice.

Also adds a simple benchmark of pooled vs unpooled of medium size buffers.

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/client/http/HttpClientIntegrationTest.java [hunk 1]
```diff
@@ -48,35 +51,134 @@
...
+            ByteBuf buf = ctx.alloc().buffer();
+            buf.writeCharSequence("pooled content", StandardCharsets.UTF_8);
+            releasedByteBuf.set(buf);
+            res.respond(HttpStatus.OK, MediaType.PLAIN_TEXT_UTF_8, new ByteBufHttpData(buf, false));
+        }
+    }
 
-        try {
+    @ClassRule
+    public static final ServerRule server = new ServerRule() {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/http/HttpClientIntegrationTest.java]: Thanks for converting to `ServerRule`!

File: core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java [hunk 2]
```diff
@@ -112,6 +117,8 @@
     @SuppressWarnings("unused")
     private volatile SubscriptionImpl subscription; // set only via subscriptionUpdater
 
+    private volatile boolean withPooledObjects;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java]: This could be a final field of `SubscriptionImpl`?

Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java]: Indeed - fixed

**Comment by trustin** [comment]
Thanks, @anuraaga !

**Pull request merged**